pkg/policy/validate.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2023 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policy
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/sigstore/policy-controller/pkg/apis/policy/v1alpha1"
+	"github.com/sigstore/policy-controller/pkg/apis/policy/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"knative.dev/pkg/apis"
+)
+
+var (
+	// ErrEmptyDocument is the error returned when no document body is
+	// specified.
+	ErrEmptyDocument = errors.New("document is required to create policy")
+
+	// ErrUnknownType is the error returned when a type contained in the policy
+	// is unrecognized.
+	ErrUnknownType = errors.New("unknown type")
+)
+
+// Validate decodes a provided YAML document containing zero or more objects
+// and performs limited validation on them.
+func Validate(ctx context.Context, document string) (warns error, err error) {
+	if len(document) == 0 {
+		return nil, ErrEmptyDocument
+	}
+
+	uol, err := Parse(ctx, document)
+	if err != nil {
+		return nil, err
+	}
+
+	for i, uo := range uol {
+		switch uo.GroupVersionKind() {
+		case v1beta1.SchemeGroupVersion.WithKind("ClusterImagePolicy"):
+			if warns, err = validate(ctx, uo, &v1beta1.ClusterImagePolicy{}); err != nil {
+				return
+			}
+
+		case v1alpha1.SchemeGroupVersion.WithKind("ClusterImagePolicy"):
+			if warns, err = validate(ctx, uo, &v1alpha1.ClusterImagePolicy{}); err != nil {
+				return
+			}
+
+		case corev1.SchemeGroupVersion.WithKind("Secret"):
```

Comment:
If the idea is that you can just cat a directory containing whatever into a single yaml, I guess I could see it, but that's not case, since if there's an unknown type we err, so yeah, just curious why the secret is special here.

Reply:
I am mainly thinking how this will start to look like assuming we start adding more things: configmaps, trustroots into the mix. That's going to make it more complicated to do more meaningful validation without starting to bleed more complexity here that will need to also be then replicated into the Compile. Secret reference support is actually a great example of that. Would the Validate unmarshal the key and make sure it's of the right type to make sure it's ok? Only to just to throw it away and redo that work in the Compile. As it stands now it doesn't do much as a validator, and actually could fail if the policy-controller is installed into a different namespace.

Not trying to be difficult here, just trying to visualize and see how this will work with more complex cases and what the division of labor is between Validate and Compile, and if Validate actually provides enough value besides making sure that the YAML can be semantically turned into objects we care about, does that make sense?

I think for CIPs it makes sense to do the Validate but for others, I am not quite convinced. At the very least, I think I'd remove the hardcoded cosign-system requirement.

Could you open a tracking issue for how much validation we really want to do besides for objects that we do not "own" (like secrets, configmaps) here vs. in Compile so we can maybe discuss it there.